### PR TITLE
kubectl: Allow []byte config fields to be set by the cli

### PR DESCRIPTION
**Pull request by petervo**
Allows []byte config fields such as 'certificate-authority-data' to be set using `kubectl config set` commands.

### CODE REVIEW

**Comment by petervo** [comment]
I should be covered by the Red Hat CLA. @eparis or @smarterclayton do i need to be added to a list?

File: pkg/kubectl/cmd/config/set.go [hunk 1]
```diff
@@ -159,6 +160,23 @@ func modifyConfig(curr reflect.Value, steps *navigationSteps, propertyValue stri
...
+			return fmt.Errorf("can't have more steps after bytes. %v", steps)
+		}
+		innerKind := actualCurrValue.Type().Elem().Kind()
+		if innerKind != reflect.Uint8 {
+			return fmt.Errorf("Unrecognized slice type. %v", innerKind)
+		}
+
+		val, err := base64.StdEncoding.DecodeString(propertyValue)
+		if err != nil {
+			val = []byte(propertyValue)
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/config/set.go]: I think you should return the error here rather than silently using the raw value.

Comment (petervo) [inline pkg/kubectl/cmd/config/set.go]: So only accept base64'd strings? Do I need to document that somewhere?

Comment (jlowdermilk) [inline pkg/kubectl/cmd/config/set.go]: The right place to document is in the [long help string](https://github.com/petervo/kubernetes/blob/byte-args/pkg/kubectl/cmd/config/set.go#L41) for the command. I'd also be fine with having a flag that controls this behavior, e.g. `--set-raw-bytes`. My concern is that silent defaulting can be surprising/confusing in edge cases.

**Comment by petervo** [comment]
Added --set-raw-bytes flag as suggested.

File: pkg/kubectl/cmd/config/set.go [hunk 2]
```diff
@@ -62,6 +66,8 @@ func NewCmdConfigSet(out io.Writer, configAccess ConfigAccess) *cobra.Command {
 		},
 	}
 
+	f := cmd.Flags().VarPF(&options.setRawBytes, "set-raw-bytes", "", "Write the given PROPERTY_VALUE directly to the binary field rather than base64 decoding first")
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/config/set.go]: Suggested help string: `When writing a []byte PROPERTY_VALUE, write the given string directly without base64 decoding.`

**Comment by jlowdermilk** [comment]
lgtm, but cc @deads2k for second pair of eyes. You'll also need to run `hack/update-generated-docs.sh`

**Comment by petervo** [comment]
Updated the doc string and ran ```hack/update-generated-docs.sh```

File: pkg/kubectl/cmd/config/set.go [hunk 3]
```diff
@@ -159,6 +171,31 @@ func modifyConfig(curr reflect.Value, steps *navigationSteps, propertyValue stri
...
+			actualCurrValue.Set(reflect.Zero(actualCurrValue.Type()))
+			return nil
+		}
+
+		if setRawBytes {
+			actualCurrValue.SetBytes([]byte(propertyValue))
+		} else {
+			val, err := base64.StdEncoding.DecodeString(propertyValue)
+			if err != nil {
+				return err
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/config/set.go]: Nit: this error is likely obtuse; should probably with `error decoding input value: %v`.

**Comment by petervo** [comment]
Updated decoding error message

File: pkg/kubectl/cmd/config/set.go [hunk 4]
```diff
@@ -159,6 +171,31 @@ func modifyConfig(curr reflect.Value, steps *navigationSteps, propertyValue stri
 		actualCurrValue.SetString(propertyValue)
 		return nil
 
+	case reflect.Slice:
+		if steps.moreStepsRemaining() {
+			return fmt.Errorf("can't have more steps after bytes. %v", steps)
+		}
+		innerKind := actualCurrValue.Type().Elem().Kind()
+		if innerKind != reflect.Uint8 {
+			return fmt.Errorf("Unrecognized slice type. %v", innerKind)
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/config/set.go]: Sorry, last nit. Error strings should not be capitalized.

**Comment by petervo** [comment]
Removed caps in error string

**Comment by jlowdermilk** [comment]
Thanks, @petervo

**Comment by deads2k** [comment]
Code looks ok, but how do I actually feed the raw bytes in as a CLI arg?  Does bash do right with `kubectl set foo $(cat my-cert-file)`?

Can you add a test to `test-cmd` feeding in something like `kubectl set foo $(cat kubectl)` or another file that contains unprintable characters?

**Comment by petervo** [comment]
The easiest way to write actual binary data is to base64 encode it.  Your example would work with say a PEM encoded certificate but not a DER one. Having the raw option is only to help avoid a step (base64 encoding) when you know it's not necessary.

**Comment by deads2k** [comment]
> The easiest way to write actual binary data is to base64 encode it. Your example would work with say a PEM encoded certificate but not a DER one. Having the raw option is only to help avoid a step (base64 encoding) when you know it's not necessary.

@petervo Can you add an example of using this flag to `test-cmd.sh`?  I realize I forgot the `--set-raw-bytes` flag in the example.

Is this for anything else in this file you'd use it for?  The cert bytes should probably be set with `kubectl config set-credentials nick-name --client-certificate=filename --embed-certs`.

@jlowdermilk I'm having some trouble understanding the use-case and how I'd be expected to use this.  I'd like to hold it just until I get that; probably just a `test-cmd.sh` example and why `kubectl config set-credentials` isn't good enough.  Code looks fine.

**Comment by petervo** [comment]
@deads2k using the existing flags, ei --client-certificate requires having the certificate present in the file on the system. So when working on a remote system, you end up with a flow like, upload the cert to a tmp location, run the command with that tmp file, and then delete the file. This is just makes it easier to change this data directly, without needing setup and cleanup steps.

There are tests in this PR that match the way the rest of kubectl config commands are tested including tests for the --set-raw-bytes flag. There are no kubectl config tests in test-cmd.sh. I'm unsure what exactly you are looking for there that isn't already tested by the tests added to config_test.go

**Comment by deads2k** [comment]
@petervo go ahead and add tests to `test-cmd` that demonstrate using these flags in a bash script.  Demonstrating with a synthetic call is good, but not something I can point a sysadmin at and say "do it like this".

**Comment by petervo** [comment]
@deads2k, ATM the only []byte fields in the config structs are certificate or key fields. I'd really rather not add openssl commands to test-cmd to generate valid cerificates. Is it ok to reference the one in test/images/porter. Or where would be the appropriate place to add a pre-generated one?

**Comment by deads2k** [comment]
> Or where would be the appropriate place to add a pre-generated one?

Pregenerated is fine.  `test/fixtures/pkg/kubectl/cmd` and subfolders is the "right" spot.

**Comment by petervo** [comment]
Added the test.

**Comment by deads2k** [comment]
Ok, doesn't look like a huge gain over `kubectl config set foo $(echo $content | base64 -)`, but I'm ok with it.

File: hack/test-cmd.sh [hunk 5]
```diff
@@ -267,6 +268,28 @@ runTests() {
   ! kubectl create
 
   #######################
+  # kubectl config set #
+  #######################
+
+  kube::log::status "Testing kubectl(${version}:config set)"
+
+  kubectl config set-cluster test-cluster --server="https://127.0.0.1:${TLS_API_PORT}"
```
Comment (petervo) [inline hack/test-cmd.sh]: Is there any way to get the value of the ip address that kube is using here. localhost and 127.0.0.1 aren't in the autogenerated certs subject-alt-names. I guess the alternative is to skip the check that this actually works. ```kubectl api-versions --cluster=test-cluster --username=none```.

Changing the service-cluster-ip-range breaks other tests.

**Comment by petervo** [comment]
Removed the line of the ```test-cmd``` test that actually checked the https connection. Because getting it to work would require generating a new certificate that has 127.0.0.1 in the alt names.

**Comment by petervo** [comment]
Rebased, only conflict was the "Auto generated by" line in the doc file. 

@deads2k or @jlowdermilk now that I had to change the PR again it needs the lgtm flag set again.

**Comment by petervo** [comment]
Flake seems to be #24265

**Comment by petervo** [comment]
@deads2k, what's the magic incantation to get the bot to get @k8s-bot to retest?

**Comment by deads2k** [comment]
> @deads2k, what's the magic incantation to get the bot to get @k8s-bot to retest?

I never learned it.  I just rebase and repush.

**Comment by petervo** [comment]
@deads2k , rebased again because of flakes. If there is a better way to this lmk.

**Comment by petervo** [comment]
rebased for flakes

**Comment by petervo** [comment]
Rebased again to disable travis tests (#23611)

**Pull request merged**